sessionization/sessionization.md

```diff
@@ -0,0 +1,597 @@
+- check the buffer for a preceding (or started simultaneously) window and all subsequent windows that intersect with the input window without having to repeatedly sort windows;
+- remove intersecting windows from the buffer;
+- and last but not least, add new windows.
+
+Sorted, tree based maps are likely the most obvious fit for above requirements.
+In the following we're going to explore an implementation based on Java's `TreeMap`, a self-balancing Red-Black tree.
+
+The aggregator will be of the following type:
+
+```java
+Aggregator<
+  // Windowed key-value pairs as input
+  WindowedValue<KV<KeyT, InputT>>,
+  // TreeMap based aggregation buffer to track accumulated inputs per session IntervalWindow
+  TreeMap<IntervalWindow, AccumT>,
+  // Collection of windowed output values, one per accumulated session
+  Collection<WindowedValue<OutputT>>>
+>
+```
+
+**Note**:
+To reduce complexity for demonstration purposes, we assume a `WindowingStrategy` with timestamp combiner `END_OF_WINDOW`.
+Support for Beam's windowing model is still partial that way. Similar to the way combine function aggregate input values, timestamp combiners aggregate input timestamps to assigne one resulting timestamp to the windowed aggregation result `WindowedValue<OutputT>`.
+In case of the default combiner `END_OF_WINDOW`, this timestamp is always the end of the session `IntervalWindow` and doesn't depend on individual input timestamps. A feature-complete implementation supporting timestamp combiners can be found on [<img src="github.png" style="height:16px;"/> apache/beam](https://github.com/apache/beam/blob/3a4d57eb8976c5f503b32d478a80b1800490f66f/runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/Aggregators.java).
+
+##### Adding values to the aggregation buffer
+
+Using the sorted tree based aggregation buffer, we can traverse the tree in order starting from the window preceding (or matching the start of) the input window if it intersects the input window, or from the following window otherwise. From there we continue to traverse the tree as long as windows intersect the input window which we continuously expand during the process. Alongside that, we have to merge corresponding accumulators.
+
+Once done, we have to remove all windows we've merged while traversing the tree. And finally add the input value to the merged accumulator (or a new empty accumulator otherwise) and insert it into the tree together with the expanded window.
+Adding the input last to the accumulator helps to avoids unnecessary allocations as we only create a new, empty accumulator if needed.
+
+The following sequence of `reduce` operations for a single key illustrates the above process. The input is highlighted in red:
+
+1. No window precedes input `7` and `[4]` doesn't intersect; create an empty accumulator `[]`, add `7` to it and insert it (into the tree).
+    ```mermaid
```

Comment:
```mermaid
    gantt
        dateFormat  Y
        axisFormat  %y
        todayMarker off
        section Key 1
        7   :crit,1, 3
        [4] :6,9
        [2] :14,15
```

Reply:
yeah at least in the generated html this diagram is not rendered